### PR TITLE
Add tests for new Rational

### LINKED ISSUE #4315: Tests for new Rational
### Describe the feature you'd like

This is pending the landing of #4284. As [requested](https://github.com/input-output-hk/plutus/pull/4284#discussion_r778775751), a port of [MLabs' extensive tests](https://github.com/Liqwid-Labs/plutus-extra/tree/master/plutus-numeric/test) for the new `Rational` upstream is a good idea. There are some outstanding questions:

* What should we do about the size tests? This has a related issue in #4314.
* We use a lot of QuickCheck-driven law tests for various Plutus type classes, but the tests in Plutus seem to be Hedgehog-based. Should these be ported too, along with their attendant scaffolding?

### Describe alternatives you've considered

Not having these tests is not a good alternative, and the port was only not done previously due to previously-described issues.

### PR DESCRIPTION
**Pull request by kozross**
This closes #4315. In particular, the following tests (all [adapted from our work at MLabs](https://github.com/Liqwid-Labs/plutus-extra/tree/master/plutus-numeric/test)) provide the following:

* Regression tests for AST sizes for all updated functionality in `Rational` (PR #4284);
* Comparison tests for sizes of `negate`, `abs` and `scale` versus generic implementations or implementations using other primitives, as appropriate;
* Checks of common laws for Plutus type classes (such as `Eq`, `Ord`, `AdditiveSemigroup` etc) for `Rational`;
* Property tests for all `Rational`-specific functionality.

We also provide some helper functionality to deal with possible coverage issues, as well as enabling function property tests for functions on `Rational`. These are currently localized to the tests which use them, but could be made more generally available if needs be.

### CODE REVIEW

**Comment by michaelpj** [comment]
Wonderful, thanks! I'll take a look later.

File: plutus-tx-plugin/test/size/Main.hs [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+-- Size testing for Tasty
+
+fitsInto :: forall (a :: Type) .
+  (Typeable a) =>
+  String -> CompiledCode a -> Integer -> TestTree
+fitsInto name cc = singleTest name . SizeTest cc
+
+data SizeTest (a :: Type) = SizeTest (CompiledCode a) Integer
+
+instance (Typeable a) => IsTest (SizeTest a) where
```
Comment (michaelpj) [inline plutus-tx-plugin/test/size/Main.hs:84]: Cute, I haven't done this style of making providers before. Is there an advantage over a function that produces an assertion?

File: plutus-tx-plugin/test/size/Main.hs [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+  run _ (SizeTest cc limit) _ = do
+    let estimate = sizePlc cc
+    let diff = limit - estimate
+    pure $ case signum diff of
+      (-1) -> testFailed $ "Exceeded limit by " <> (show . abs $ diff)
+      0    -> testPassed $ "AST size: " <> show limit
+      _    -> testPassed $ "Remaining budget: " <> show diff
```
Comment (michaelpj) [inline plutus-tx-plugin/test/size/Main.hs:91]: (just because "budget" is a bit of an overloaded word)

```suggestion
      _    -> testPassed $ "Remaining headroom: " <> show diff
```

File: plutus-tx-plugin/test/size/Main.hs [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+                                "Exceeded by: " <> show diff
+
+renderEstimates :: (String, Integer) -> (String, Integer) -> String
+renderEstimates (tName, tEstimate) (mName, mEstimate) =
+  "Target: " <> tName <> "; size " <> show tEstimate <> "\n" <>
+  "Measured: " <> mName <> "; size " <> show mEstimate <> "\n"
+
+renderExcess :: (String, Integer) -> (String, Integer) -> Integer -> String
+renderExcess tData mData diff = renderEstimates tData mData <>
+                                "Remaining slack: " <> show diff
```
Comment (michaelpj) [inline plutus-tx-plugin/test/size/Main.hs:124]:
```suggestion
                                "Remaining headroom: " <> show diff
```

File: plutus-tx/test/laws/Main.hs [hunk 4]
```diff
@@ -0,0 +1,533 @@
...
+  den <- Gen.integral . Range.linearFrom 100 0 $ 200
+  pure . Ratio.unsafeRatio num $ den
+
+genInteger :: Gen Integer
+genInteger = Gen.integral . Range.linearFrom 0 (-100) $ 100
+
+genIntegerPos :: Gen Integer
+genIntegerPos = Gen.integral . Range.linearFrom 50 1 $ 100
+
+-- This is a hack to ensure coverage.
```
Comment (michaelpj) [inline plutus-tx/test/laws/Main.hs:497]: Can you explain the hack a bit more for those of us not in the know?

File: plutus-tx/test/laws/Main.hs [hunk 5]
```diff
@@ -0,0 +1,533 @@
...
+  else success
+
+propEqSub :: Property
+propEqSub = property $ do
+  f <- forAllFn . fnWith varyRational $ genInteger
+  ent <- forAllWith ppShow (entangled genRational)
+  cover 45 "identical" . isDefinitelyIdentical $ ent
+  cover 45 "possibly different" . not . isDefinitelyIdentical $ ent
+  let (x, y) = collapse ent
+  if x Plutus.== y
```
Comment (michaelpj) [inline plutus-tx/test/laws/Main.hs:130]: same question

Comment (michaelpj) [reply plutus-tx/test/laws/Main.hs:130]: and in lots of places below

**Review by michaelpj — COMMENTED** [review]
Thank you very much for these thorough tests! Mostly just organizational comments.

**Comment by michaelpj** [comment]
You'll need to run `updateMaterialized` at some point to make the CI happy, since you've added some modules and components.

**Comment by kwxm** [comment]
I'll take a closer look at this early tomorrow; after a quick glance it looks very good though!

File: plutus-tx-plugin/test/size/Main.hs (hunk 1 above)
Comment (kozross) [reply plutus-tx-plugin/test/size/Main.hs:84]: This is the correct method for providing new runners to `tasty`. This also gives me full control over output, as well as the ability to add any `tasty` options I see fit.

File: plutus-tx/test/laws/Main.hs [hunk 6]
```diff
@@ -0,0 +1,533 @@
...
+import Hedgehog.Range qualified as Range
+import PlutusTx.IsData.Class (fromBuiltinData, toBuiltinData, unsafeFromBuiltinData)
+import PlutusTx.Prelude qualified as Plutus
+import PlutusTx.Ratio qualified as Ratio
+import Test.Tasty (defaultMain, localOption, testGroup)
+import Test.Tasty.Hedgehog (HedgehogTestLimit (HedgehogTestLimit), testProperty)
+import Text.Show.Pretty (ppShow)
+
+main :: IO ()
+main = defaultMain . localOption (HedgehogTestLimit . Just $ 1000) . testGroup "Rational" $ [
```
[the remark this answers is not in the document]
Comment (kozross) [reply plutus-tx/test/laws/Main.hs:27]: Because 100 tests is too few to indicate anything, and probabilistic 'swing' means you fail coverage half the time.

File: plutus-tx/test/laws/Main.hs (hunk 4 above)
Comment (kozross) [reply plutus-tx/test/laws/Main.hs:497]: This is quite complex, but I'll try and keep it brief. Consider the symmetry property of a binary relation `R`. We have the property that 'if `R(x, y)`, then `R(y,x)`. If some `x` and `y` are not related, we actually don't care what the outcome is.

When testing the symmetry of _equality_, if we generate both `x` and `y` independently, there is a high probability that we will generate non-equal data. Therefore, if we run the test 'blindly', we're not actually testing anything: most of the time, the test will trivially pass because its precondition is not satisfied by the generated data!

`Entangled` (and `Entangled3`) essentially 'hack' the generator to ensure that half the time we generate the same thing twice (or three times), and the other half the time, we generate normally. Combined with liberal use of `cover`, we can ensure that we cover all possibilities. In fact, `cover` gets used in several places, as this ensures that we don't bias our testing to any particular set of preconditions.

File: plutus-tx/plutus-tx.cabal [hunk 7]
```diff
@@ -117,3 +117,18 @@ test-suite plutus-tx-test
         tasty-hunit -any,
         serialise -any,
         cborg -any
+
+test-suite laws
```
[the remark this answers is not in the document]
Comment (kozross) [reply plutus-tx/plutus-tx.cabal:121]: Being able to easily run these independently, mainly. I'm happy to combine these with the main suite.

File: plutus-tx-plugin/plutus-tx-plugin.cabal [hunk 8]
```diff
@@ -130,3 +129,15 @@ test-suite plutus-tx-tests
         containers -any
     -- NOTE: -g makes the plugin give better errors
     ghc-options: -g
+
+test-suite size
```
[the remark this answers is not in the document]
Comment (kozross) [reply plutus-tx-plugin/plutus-tx-plugin.cabal:133]: This is a regression test specifically for size, which I think benefits from being independently-runnable easily.

File: cabal.project [hunk 9]
```diff
@@ -21,7 +21,7 @@ tests: true
 benchmarks: true
 
 -- The only sensible test display option
-test-show-details: streaming
+test-show-details: direct
```
[the remark this answers is not in the document]
Comment (michaelpj) [reply cabal.project:24]: Does it? In what way? I don't mind taking a change for this, I'd just like some justification rather than a random diff in the middle of another PR :)

File: plutus-tx-plugin/plutus-tx-plugin.cabal (hunk 8 above)
[the remark this answers is not in the document]
Comment (michaelpj) [reply plutus-tx-plugin/plutus-tx-plugin.cabal:133]: Isn't that what tasty filters are for?

File: plutus-tx/test/laws/Main.hs (hunk 6 above)
[the remark this answers is not in the document]
Comment (michaelpj) [reply plutus-tx/test/laws/Main.hs:27]: in the code!

File: plutus-tx/test/laws/Main.hs (hunk 4 above)
Comment (michaelpj) [reply plutus-tx/test/laws/Main.hs:497]: write it in the code!

File: cabal.project (hunk 9 above)
[the remark this answers is not in the document]
Comment (kozross) [reply cabal.project:24]: `streaming` doesn't colourize output, and `tasty` (or at least, some of its runners) provide colourized diffs. It also gives (using _any_ runner) a clear indication of passed tests (which are green) and failed ones (which are red).

File: plutus-tx-plugin/plutus-tx-plugin.cabal (hunk 8 above)
[the remark this answers is not in the document]
Comment (kozross) [reply plutus-tx-plugin/plutus-tx-plugin.cabal:133]: These are not easy to use. `cabal new-test size` is considerably less awkward than `cabal new-test --test-options="[insert escaping hell here]"`. It's also pretty bad to have to play guesswork with the pseudo-awk syntax that `tasty` provides for filtering.

File: cabal.project (hunk 9 above)
[the remark this answers is not in the document]
Comment (kwxm) [reply cabal.project:24]: Hmmm.  I'm sure I used to see colourised test results, but indeed with `streaming` I don't. The `streaming` option seems to have been there for about six months though.

**Review by kwxm — COMMENTED** [review]
This is great, and it's refreshing to see code from a new contributor!  I have no real complaints apart from the issue about coverage failures.

One extra comment: it's an important invariant of the implementation that the results of all the operations have a positive denominator and coprime numerator and denominator.  You have a standalone test for the former but not the latter.  Could you do something along the lines of building checks of the invariants into `===`?  That would give you thousands of checks of the invariants quite cheaply.

More generally, this is all testing the Haskell implementation of the operations, and I wonder if we should also test the compiled Plutus versions, although that would be more a test of the compiler than the library.  We do have [some tests](https://github.com/input-output-hk/plutus/blob/master/plutus-benchmark/nofib/test/Spec.hs) that Plutus Core code does the same as the Haskell it's compiled from, but maybe we should test library functions more aggressively.  I'm definitely not suggesting that you should do this!  Maybe it's something that we at IOHK should think about though.

File: cabal.project (hunk 9 above)
[the remark this answers is not in the document]
Comment (michaelpj) [reply cabal.project:24]: Okay, good to know. Can you add a comment saying "this is the only mode that allows colorized output from tasty"

File: plutus-tx-plugin/plutus-tx-plugin.cabal (hunk 8 above)
[the remark this answers is not in the document]
Comment (michaelpj) [reply plutus-tx-plugin/plutus-tx-plugin.cabal:133]: Okay

File: plutus-tx/plutus-tx.cabal (hunk 7 above)
[the remark this answers is not in the document]
Comment (michaelpj) [reply plutus-tx/plutus-tx.cabal:121]: These ones I don't see much reason to separate, since they're functional tests.

File: plutus-tx/test/laws/Main.hs [hunk 10]
```diff
@@ -0,0 +1,533 @@
...
+separate r = (Ratio.numerator r, Ratio.denominator r)
+
+genRational :: Gen Plutus.Rational
+genRational = do
+  num <- genInteger
+  den <- Gen.integral . Range.linearFrom 100 0 $ 200
+  pure . Ratio.unsafeRatio num $ den
+
+genInteger :: Gen Integer
+genInteger = Gen.integral . Range.linearFrom 0 (-100) $ 100
```
[the remark this answers is not in the document]
Comment (kozross) [reply plutus-tx/test/laws/Main.hs:492]: I think honestly going much larger than this has few advantages. I can't think of many cases where our tests would uncover something with larger numbers than can't be found with smaller ones, and as you correctly observed, the default size is 30 anyway. I feel fairly confident this range is about right.

**Comment by kozross** [comment]
> One extra comment: it's an important invariant of the implementation that the results of all the operations have a positive denominator and coprime numerator and denominator. You have a standalone test for the former but not the latter. Could you do something along the lines of building checks of the invariants into `===`? That would give you thousands of invariant checks quite cheaply.

This is a valid point. I don't use `===` only for comparing `Rational`s, so I'd need both a specialized function (for `Rational`s) _and_ `===`, but it's a good strategy.
 
> More generally, this is all testing the Haskell implementation of the operations, and I wonder if we should also test the compiled Plutus versions, although that would be more a test of the compiler than the library. We do have [some tests](https://github.com/input-output-hk/plutus/blob/master/plutus-benchmark/nofib/test/Spec.hs) that Plutus Core code does the same as the Haskell it's compiled from, but maybe we should test library functions more aggressively. I'm definitely not suggesting that you should do this! Maybe it's something that we at IOHK should think about though.

This is... notably more difficult. I'm certainly not against this (I'm pretty much _the_ biggest stickler for testing you'll ever find), but I would need @michaelpj to approve this, and it would _definitely_ need to be a separate PR to this one. FWIW, the tests you link to are very much of the 'unit test' stripe, which I have little confidence in: I would much prefer something like a property-based approach, but this would require more work.

**Comment by kwxm** [comment]
> This is... notably more difficult. I'm certainly not against this (I'm pretty much _the_ biggest stickler for testing you'll ever find), but I would need @michaelpj to approve this, and it would _definitely_ need to be a separate PR to this one. FWIW, the tests you link to are very much of the 'unit test' stripe, which I have little confidence in: I would much prefer something like a property-based approach, but this would require more work.

It's just something that occurred to me as I was looking at this; as I say, I'm not suggesting that it's something you should tackle.  I can open a ticket to remind us to think about this when/if we have some spare time.  I agree that property-based tests would be better.  In fact I don't think we've ever actually had any problems in the plugin that didn't show up very quickly with our current tests, but more tests are always reassuring.  Anyway, just ignore the comment above.

File: plutus-tx/test/laws/Main.hs [hunk 11]
```diff
@@ -0,0 +1,533 @@
...
+
+propRatioSelf :: Property
+propRatioSelf = property $ do
+  x <- forAllWith ppShow . Gen.filter (/= Plutus.zero) $ genInteger
+  Ratio.ratio x x === Just Plutus.one
+
+propRatioSign :: Property
+propRatioSign = property $ do
+  (n, d) <- forAllWith ppShow go
+  cover 30 "zero numerator" $ n == 0
```
[the remark this answers is not in the document]
Comment (kwxm) [reply plutus-tx/test/laws/Main.hs:325]: > FWIW, I've found that current settings give such failures rarely, but I'm happy to adjust to make it even less likely.

The problem is that any errors in the tests will cause CI to fail, so we might get random failures in completely unrelated PRs because of this, and when that happens someone has to restart the CI job for the PR manually.  If it happened one time in a thousand that _might_ be bearable, but still not ideal.  I just re-ran the tests another 100 times and this time got seven coverage failures, which is definitely too much.

[the remark this answers is not in the document]
Comment (kozross) [reply plutus-tx/test/laws/Main.hs:325]: I'll do similar experiments locally with more tests and lower coverage boundaries, using 1-in-a-thousand as the desired target. It'll take a while, but I can definitely bring it within that bound.

**Comment by kozross** [comment]
I tried to do `updateMaterialized`, but now the MinGW builds are blowing up.

**Comment by kozross** [comment]
@kwxm My latest commit increases the test count for any tests involving `cover` to 8000. This means that, even if I run 2000 tests, I don't see any failures. Feel free to test this locally to be sure. This doesn't increase test time much (overall, the tests from `laws` take about 1.5 seconds on my machine), so if you're happy, I guess we can consider this solved?

**Comment by kwxm** [comment]
> @kwxm My latest commit increases the test count for any tests involving `cover` to 8000. This means that, even if I run 2000 tests, I don't see any failures. Feel free to test this locally to be sure. This doesn't increase test time much (overall, the tests from `laws` take about 1.5 seconds on my machine), so if you're happy, I guess we can consider this solved?

Yep.  I ran the tests 1000 times with no coverage failures, so that seems to have sorted it out.  Thanks! I think I'd be happy to merge this now, although there's a mystery build failure that I don't recall having seen before.

File: plutus-tx/test/laws/Main.hs [hunk 12]
```diff
@@ -0,0 +1,582 @@
...
+  pure . Ratio.unsafeRatio num $ den
+
+genInteger :: Gen Integer
+genInteger = Gen.integral . Range.linearFrom 0 (-100) $ 100
+
+genIntegerPos :: Gen Integer
+genIntegerPos = Gen.integral . Range.linearFrom 50 1 $ 100
+
+-- This is a hack to ensure coverage.
+--
```
Comment (kwxm) [inline plutus-tx/test/laws/Main.hs:510]: You could maybe use `{- ... -}` here, but no big deal.  Good comment though.

File: plutus-tx/test/laws/Main.hs (hunk 11 above)
[the remark this answers is not in the document]
Comment (michaelpj) [reply plutus-tx/test/laws/Main.hs:337]: We're definitely not asking you to fix Hedgehog :)

File: plutus-tx/test/laws/Main.hs (hunk 10 above)
[the remark this answers is not in the document]
Comment (kwxm) [reply plutus-tx/test/laws/Main.hs:504]: OK.

**Review by kwxm — APPROVED** [review]
All looks good.

**Comment by michaelpj** [comment]
You need to delete the unused `plutus-tx-plugin` dep from the size test suite, and then rerun `updateMaterialized`. I'd be happy to do this for you, but you would need to tick the "allow edits from maintainers" box on the PR so I have push access to the PR branch on your fork.

File: plutus-tx/test/laws/Main.hs (hunk 12 above)
Comment (kozross) [reply plutus-tx/test/laws/Main.hs:510]: I prefer `--`-style multilines, not least for all because they're easier to deal with in Neovim.

**Comment by kozross** [comment]
@kwxm @michaelpj The latest commit should address all outstanding issues. In particular, I have put the laws tests into separate suites, and merged them into the test suite for `plutus-tx`.

@michaelpj I tried following your suggestion for deleting the unused dep and rerunning `updateMaterialized`, but then the size tests no longer compile.

**Comment by michaelpj** [comment]
Thanks! I don't want to put you through another roundtrip of fixing this stuff, so I'm just going to fix it myself for you and push that up.

**Comment by michaelpj** [comment]
Merged in https://github.com/input-output-hk/plutus/pull/4340. Thanks again!
